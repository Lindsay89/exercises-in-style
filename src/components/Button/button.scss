@import "../../utils/variables.scss";

.eis-button {
  padding: 0.5rem;
  display: flex;
  align-items: center;
  color: $text-color;
  &.primary {
    background-color: $primary-color;
    color: white;
  }
}
