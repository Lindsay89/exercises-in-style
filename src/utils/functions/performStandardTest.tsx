import React from "react";
import { render } from "@testing-library/react";
import { DefaultProps } from "../types/defaultProps";

const performStandardTest = (
  Component: React.FunctionComponent<any>,
  defaultProps: DefaultProps,
  defaultClassName: string
) => {
  const dataTestId = defaultProps["data-testid"] as string;
  it("should render without explode", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).id).toMatchSnapshot();
  });

  it('should accept an "id" prop', () => {
    const { getByTestId } = render(<Component {...defaultProps} id="foo-id" />);

    expect(getByTestId(dataTestId).id).toBe("foo-id");
  });

  it("should have default className", () => {
    const { getByTestId } = render(<Component {...defaultProps} />);

    expect(getByTestId(dataTestId).classList?.contains(defaultClassName)).toBe(
      true
    );
  });

  it("should allow adding custom classes", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} className="bar" />
    );

    expect(getByTestId(dataTestId).classList.contains("bar")).toBe(true);
  });

  it("should allow to define custom style", () => {
    const { getByTestId } = render(
      <Component {...defaultProps} style={{ margin: "30px" }} />
    );

    expect(getByTestId(dataTestId).getAttribute("style")).toContain("margin");
  });
};
export default performStandardTest;
