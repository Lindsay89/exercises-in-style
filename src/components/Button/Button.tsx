import React, { ButtonHTMLAttributes } from "react";
import classNames from "classnames";
import { DefaultProps } from "../../utils/types";
import * as FeatherIcon from "react-feather";
import { SizeStandard } from "../../utils/types/size";

// styles
import "./button.scss";

// Required Props
interface ButtonRequiredProps {
  // The function to perform on button click
  onClick: () => void;
}

// Optional Props
interface ButtonOptionalProps extends DefaultProps {
  // The button children
  children?: React.ReactNode;
  // The button' size
  size?: SizeStandard;
  // The button's type
  type?: "primary" | "secondary" | "ghost";
  rounded?: boolean;
  outlined?: boolean;
  icon?: keyof typeof FeatherIcon;
}

// Combined required and optional props to build the full prop interface
export interface ButtonProps
  extends ButtonRequiredProps,
    ButtonOptionalProps,
    Omit<ButtonHTMLAttributes<HTMLButtonElement>, "type" | "onClick"> {}

// use the optional prop interface to define the default props
const defaultProps: ButtonOptionalProps = {
  "data-testid": "eis-button",
  size: "medium",
  type: "primary",
  rounded: false,
  outlined: false,
};

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      children,
      onClick,
      size,
      type,
      rounded,
      outlined,
      icon,
      ...rest
    },
    ref
  ) => {
    const classList = classNames(
      "eis-button",
      size,
      type,
      { rounded },
      { outlined },
      className
    );
    const Icon = icon && FeatherIcon[icon];

    return (
      <button ref={ref} className={classList} onClick={onClick} {...rest}>
        {children}
        {Icon && <Icon />}
      </button>
    );
  }
);

Button.defaultProps = defaultProps as Partial<ButtonOptionalProps>;

export default React.memo(Button);
