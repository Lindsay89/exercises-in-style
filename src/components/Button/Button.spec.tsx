import React from "react";
import Button from "./Button";
import performStandardTest from "../../utils/functions/performStandardTest";
import { fireEvent, getByTestId, render, screen } from "@testing-library/react";

const defaultProps = {
  "data-testid": "eis-button",
  onClick: jest.fn(),
};

describe("Button Component", () => {
  performStandardTest(Button, defaultProps, "eis-button");

  it("should perform a button click when function is provided", () => {
    const onClickFn = jest.fn();
    render(<Button onClick={onClickFn}>click</Button>);

    const EisButton = screen.getByRole("button");
    fireEvent.click(EisButton);
    expect(onClickFn).toHaveBeenCalled();
  });

  it("should render in the correct size when provided", () => {
    const { rerender } = render(
      <Button {...defaultProps} size="small">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("small");

    rerender(
      <Button {...defaultProps} size="medium">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("medium");

    rerender(
      <Button {...defaultProps} size="large">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("large");
  });

  it("should render in the correct type when provided", () => {
    const { rerender } = render(
      <Button {...defaultProps} type="primary">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("primary");

    rerender(
      <Button {...defaultProps} type="secondary">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("secondary");

    rerender(
      <Button {...defaultProps} type="ghost">
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("ghost");
  });
  it("should be rounded by default", () => {
    render(<Button {...defaultProps}>click</Button>);

    expect(screen.getByRole("button")).toHaveClass("rounded");
  });
  it("should be outlined if outlined prop is provided", () => {
    render(
      <Button {...defaultProps} outlined>
        click
      </Button>
    );

    expect(screen.getByRole("button")).toHaveClass("outlined");
  });
});
