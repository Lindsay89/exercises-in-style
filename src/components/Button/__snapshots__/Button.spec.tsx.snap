// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component should render without explode 1`] = `
<DocumentFragment>
  <button
    class="eis-button"
    data-testid="eis-button"
  />
</DocumentFragment>
`;

exports[`Button Component should render without exploding 1`] = `
<button
  class="eis-button"
  data-testid="eis-button"
/>
`;
